### ISSUE TITLE
Add some getter for the "open" property of each node

### ISSUE THREAD

**Issue by GermanJablo** [brimdata/react-arborist#207]
The `open` property (well, `collapsed` actually) is part of my controlled data.

I want it to be like this so that the open or closed state is persistent.

I've tried using the onToggle property to keep my open state in sync with react-arborist's open, and it seems to work fine. The problem is that with the initial state they become desynchronized. `openByDefault` to true or false would not solve my problem, because some nodes should be able to be initialized open and others closed.

Two ideas have occurred to me:

- Add an `openAccessor` or `getOpen` prop that works as a getter and use `onToggle` as a setter.
- Add two properties `getOpen` and `setOpen`. It probably makes sense to deprecate `onToggle` in this case.

Something I don't like about `onToggle` is that instead of passing the node, it passes the id, and I have to do a lookup of my data to find it and access it.

**Comment by jameskerr** [comment]
You're right. This is not ideal. I am considering changing the api to make this easier in the next major version. Glad you were at least able to create a workaround.

**Comment by vbynda-gresham** [comment]
> Edit: I just discovered the `initialOpenState` property that I had overlooked. With this I have been able to solve it, although it is a bit of a boilerplate since I have to go through my entire data structure and form an `OpenMap`, not ideal.

Could you show a snippet how you solved it, i think i need to solve the same problem.

**Comment by cmingxu** [comment]
any update on this ?

**Comment by sauron918** [comment]
I would be also appreciated for the change the API to: `onToggle?: (node: NodeApi<T>) => void`

**Comment by TrevorBurnham** [comment]
A few APIs now cover persisting open/closed state:

- `node.isOpen` / `node.isClosed` getters read the state per node.
- `initialOpenState` (which you found in your edit) seeds the initial open map, including a mix of open and closed nodes.
- `onToggle(id)` fires on changes; you can resolve the node with `tree.get(id)` inside it.

Between `initialOpenState` for seeding and `onToggle` for syncing, the open state can be fully persisted. Closing as covered — please reopen if the id-vs-node ergonomics of `onToggle` are still a blocker and we can look at passing the node.
